src/main/scala/viper/server/core/ViperCache.scala

```diff
@@ -26,69 +32,67 @@ object ViperCache extends Cache {
     _backendSpecificCache = backendSpecificCache
     _logger = logger
-    resetCache()
+
+    _cacheFile = cacheFile.map(file => new java.io.File(file))
+
+    if(_cacheFile.isDefined) {
+      logger.trace("Trying to initializing cache with file {}", cacheFile.get)
+      if(_cacheFile.get.exists() && _cacheFile.get.canRead) {
+        try{
+          implicit val formats: Formats = DefaultFormats.withHints(ViperCacheHelper.cacheEntryHints)
+
+          _cache = read[Cache](Files.readString(_cacheFile.get.toPath))
+          logger.trace("Successfully read cache from file {}", cacheFile.get)
+        } catch {
+          case e: Throwable =>
+            logger.warn("Reading of cache file " + cacheFile.get + "  failed, error is: " + e.getMessage)
+            logger.debug("Error thrown: {}", e)
+        }
+      } else {
+        _logger.info("Cache file " + cacheFile.get + " not found, starting with empty cache")
+      }
+    } else {
+      logger.debug("No cache file specified, starting with empty cache")
+      resetCache()
+    }
   }
 
   def applyCache(
         backendName: String,
         file: String,
         p: Program): (Program, List[CacheResult]) = {
-
-    val file_key = getKey(backendName, file)
+    val file_key = getKey(file = file, backendName = backendName)
     val cacheable_ast = ViperAst(p)
     val (output_ast, cache_entries) = super.retrieve(file_key, cacheable_ast)
     val output_prog = output_ast.asInstanceOf[ViperAst].p
 
+    implicit val formats: Formats = DefaultFormats.withHints(ViperCacheHelper.errorNodeHints(p, file_key))
+
     val ver_results = cache_entries.map(ce => {
-      val concerning_method: Method = ce.concerning.asInstanceOf[ViperMethod].m
-      val content = ce.content.asInstanceOf[ViperCacheContent]
-      val ver_errors = updateErrorLocation(file_key, p, concerning_method, content)
-      CacheResult(concerning_method, ver_errors)
-    })
+        val concerning_method = p.methods.find(method => method.entityHash == ce.concerningHash).get
+        try {
+          // Try to deserialize content of the cache entry
+          val content = read[ViperCacheContent](ce.content.asInstanceOf[SerializedViperCacheContent].content)
+          logger.trace("Got a cache hit for method " + concerning_method.name)
+          CacheResult(concerning_method, content.errors)
+        } catch {
+          case e: Throwable =>
+            // In case parsing of the cache entry fails, abort caching & evict cache entries for this file, since hey might come from an unsupported version
+            logger.warn("Parsing of CacheEntry for method " + concerning_method.name + " failed, error is:" + e.getMessage)
+            logger.debug("{}", e)
+            super.forgetFile(file_key)
+            return(p, List())
+        }
+    }).filter(e => e != null)
 
     (output_prog, ver_results)
   }
 
-  private def updateErrorLocation(
```

Comment:
Exactly, my goal was to not store any unnecessary information in the cache, we use the updated AST nodes passed with the program AST to rebuild the Verification Error objects. That's also why we serialize/deserialize in two steps because we actually need the new Program AST to resolve ErrorNodes